Need help adjusting scoping for new grammar constructs

I have extended the EquationStatement grammar to
1) have a "where" clause in which variables can be defined and in which assumptions can be stated, and
2) have a "return" statement that allows the values returned by an equation to be explicitly stated.
Note that a previous grammar change allows Equations to return multiple values, in which case a list syntax allows specifying to what variables the return values are assigned.
Scoping needs to be enhanced to do the following.
a) If a variable is defined in the "where" clause and used in the body, the definition is the reference in the "where" clause.
b) If a variable is referenced in the body and referenced in the "return" statement, the definition is the reference in the body.

These changes are pushed to branch [`AWC-ServicesConfig`](https://github.com/crapo/sadlos2/tree/AWC-ServicesConfig).

Here's an example .sadl file content:
```
uri "http://sadl.org/MinimalExample.sadl" alias mex.
 
 External unitResolver(string u, ...) returns string: "http://sadl.org/unitSelector". 
 External derivative(ScientificConcept numerator, ScientificConcept denominator, int n) returns decimal, string: 
 	"http://sadl.org/derivative".
 
 Mass is a type of UnittedQuantity.
 Velocity is a type of UnittedQuantity.
 Momentum is a type of UnittedQuantity.
 Force is a type of UnittedQuantity.
 Time is a type of UnittedQuantity.
 
 Equation newtons2ndLaw 
 	(note "Force is equal to the derivative of momentum with respect to time.") 
 	(Mass m, Velocity v) returns Force: 
 	a Force f with ^value fv, with unit fu
 	
 	return f
 	
 	where mv is a Momentum with ^value (^value of m * ^value of v), with unit unitResolver("*", unit of m, unit of v) and
 		[fv,fu] = derivative(mv, Time, 1).
 ```

@crapo, the example you have provided has compiler errors. I created a new project with SADL (Eclipse) IDE, added a new `x.sadl` file and copied over the [example from above](
https://github.com/crapo/sadlos2/issues/344#issue-382432616).

```
SLF4J: Class path contains multiple SLF4J bindings.
SLF4J: Found binding in [bundleresource://385.fwk789219251:22/org/slf4j/impl/StaticLoggerBinder.class]
SLF4J: Found binding in [bundleresource://385.fwk789219251:48/org/slf4j/impl/StaticLoggerBinder.class]
SLF4J: See http://www.slf4j.org/codes.html#multiple_bindings for an explanation.
SLF4J: Actual binding is of type [org.slf4j.impl.Log4jLoggerFactory]

!ENTRY org.apache.log4j 4 0 2018-11-27 12:47:54.876
!MESSAGE org.eclipse.xtext.validation.ResourceValidatorImpl  - java.io.FileNotFoundException: /Users/akos.kitta/dev/wss/theia-sadl/runtime-configuration-assert/test_project/ImplicitModel/SadlImplicitModel.sadl (Permission denied)

!STACK 0
org.eclipse.emf.common.util.WrappedException: java.io.FileNotFoundException: /Users/akos.kitta/dev/wss/theia-sadl/runtime-configuration-assert/test_project/ImplicitModel/SadlImplicitModel.sadl (Permission denied)
	at org.eclipse.xtext.util.Files.writeStringIntoFile(Files.java:177)
	at com.ge.research.sadl.processing.ISadlImplicitModelContentProvider.createImplicitModel(ISadlImplicitModelContentProvider.java:165)
	at com.ge.research.sadl.processing.ISadlImplicitModelContentProvider.createImplicitModel(ISadlImplicitModelContentProvider.java:180)
	at com.ge.research.sadl.processing.SadlModelProcessor.createSadlImplicitModel(SadlModelProcessor.java:667)
	at com.ge.research.sadl.jena.JenaBasedSadlModelProcessor.checkImplicitSadlModelExistence(JenaBasedSadlModelProcessor.java:11035)
	at com.ge.research.sadl.jena.JenaBasedSadlModelProcessor.addImplicitSadlModelImportToJenaModel(JenaBasedSadlModelProcessor.java:1391)
	at com.ge.research.sadl.jena.JenaBasedSadlModelProcessor.onValidate(JenaBasedSadlModelProcessor.java:1016)
	at com.ge.research.sadl.validation.ResourceValidator.doValidate(ResourceValidator.java:76)
	at com.ge.research.sadl.validation.ResourceValidator.validate(ResourceValidator.java:72)
	at org.eclipse.xtext.validation.ResourceValidatorImpl.validate(ResourceValidatorImpl.java:90)
	at com.ge.research.sadl.validation.ResourceValidator.lambda$0(ResourceValidator.java:56)
	at org.eclipse.xtext.util.OnChangeEvictingCache.get(OnChangeEvictingCache.java:76)
	at com.ge.research.sadl.validation.ResourceValidator.validate(ResourceValidator.java:58)
	at org.eclipse.xtext.ui.editor.validation.ValidationJob$1.exec(ValidationJob.java:91)
	at org.eclipse.xtext.ui.editor.validation.ValidationJob$1.exec(ValidationJob.java:1)
	at org.eclipse.xtext.util.concurrent.CancelableUnitOfWork.exec(CancelableUnitOfWork.java:26)
	at org.eclipse.xtext.resource.OutdatedStateManager.exec(OutdatedStateManager.java:91)
	at org.eclipse.xtext.ui.editor.model.XtextDocument$XtextDocumentLocker.internalReadOnly(XtextDocument.java:527)
	at org.eclipse.xtext.ui.editor.model.XtextDocument$XtextDocumentLocker.readOnly(XtextDocument.java:499)
	at org.eclipse.xtext.ui.editor.model.XtextDocument.readOnly(XtextDocument.java:137)
	at org.eclipse.xtext.ui.editor.validation.ValidationJob.createIssues(ValidationJob.java:86)
	at org.eclipse.xtext.ui.editor.validation.ValidationJob.run(ValidationJob.java:67)
	at org.eclipse.core.internal.jobs.Worker.run(Worker.java:53)
Caused by: java.io.FileNotFoundException: /Users/akos.kitta/dev/wss/theia-sadl/runtime-configuration-assert/test_project/ImplicitModel/SadlImplicitModel.sadl (Permission denied)
	at java.io.FileOutputStream.open0(Native Method)
	at java.io.FileOutputStream.open(FileOutputStream.java:270)
	at java.io.FileOutputStream.<init>(FileOutputStream.java:213)
	at com.google.common.io.Files$FileByteSink.openStream(Files.java:245)
	at com.google.common.io.Files$FileByteSink.openStream(Files.java:233)
	at com.google.common.io.ByteSink.write(ByteSink.java:111)
	at com.google.common.io.Files.write(Files.java:415)
	at org.eclipse.xtext.util.Files.writeStringIntoFile(Files.java:175)
	... 22 more
java.lang.NullPointerException
	at com.ge.research.sadl.model.gp.Equation.toFullyQualifiedString(Equation.java:256)
	at com.ge.research.sadl.jena.JenaBasedSadlModelProcessor.createExternalEquation(JenaBasedSadlModelProcessor.java:2968)
	at com.ge.research.sadl.jena.JenaBasedSadlModelProcessor.processStatement(JenaBasedSadlModelProcessor.java:2895)
	at com.ge.research.sadl.jena.JenaBasedSadlModelProcessor.processModelElement(JenaBasedSadlModelProcessor.java:1194)
	at com.ge.research.sadl.jena.JenaBasedSadlModelProcessor.onValidate(JenaBasedSadlModelProcessor.java:1078)
	at com.ge.research.sadl.validation.ResourceValidator.doValidate(ResourceValidator.java:76)
	at com.ge.research.sadl.validation.ResourceValidator.validate(ResourceValidator.java:72)
	at org.eclipse.xtext.validation.ResourceValidatorImpl.validate(ResourceValidatorImpl.java:90)
	at com.ge.research.sadl.validation.ResourceValidator.lambda$0(ResourceValidator.java:56)
	at org.eclipse.xtext.util.OnChangeEvictingCache.get(OnChangeEvictingCache.java:76)
	at com.ge.research.sadl.validation.ResourceValidator.validate(ResourceValidator.java:58)
	at org.eclipse.xtext.ui.validation.DefaultResourceUIValidatorExtension.addMarkers(DefaultResourceUIValidatorExtension.java:60)
	at org.eclipse.xtext.ui.validation.DefaultResourceUIValidatorExtension.updateValidationMarkers(DefaultResourceUIValidatorExtension.java:46)
	at org.eclipse.xtext.builder.builderState.MarkerUpdaterImpl.processDelta(MarkerUpdaterImpl.java:93)
	at org.eclipse.xtext.builder.builderState.MarkerUpdaterImpl.updateMarkers(MarkerUpdaterImpl.java:63)
	at org.eclipse.xtext.builder.builderState.AbstractBuilderState.updateMarkers(AbstractBuilderState.java:82)
	at org.eclipse.xtext.builder.clustering.ClusteringBuilderState.doUpdate(ClusteringBuilderState.java:283)
	at org.eclipse.xtext.builder.builderState.AbstractBuilderState.update(AbstractBuilderState.java:116)
	at org.eclipse.xtext.builder.impl.XtextBuilder.doBuild(XtextBuilder.java:288)
	at org.eclipse.xtext.builder.impl.XtextBuilder.incrementalBuild(XtextBuilder.java:268)
	at org.eclipse.xtext.builder.impl.XtextBuilder.build(XtextBuilder.java:162)
	at org.eclipse.core.internal.events.BuildManager$2.run(BuildManager.java:733)
	at org.eclipse.core.runtime.SafeRunner.run(SafeRunner.java:42)
	at org.eclipse.core.internal.events.BuildManager.basicBuild(BuildManager.java:206)
	at org.eclipse.core.internal.events.BuildManager.basicBuild(BuildManager.java:246)
	at org.eclipse.core.internal.events.BuildManager$1.run(BuildManager.java:299)
	at org.eclipse.core.runtime.SafeRunner.run(SafeRunner.java:42)
	at org.eclipse.core.internal.events.BuildManager.basicBuild(BuildManager.java:302)
	at org.eclipse.core.internal.events.BuildManager.basicBuildLoop(BuildManager.java:358)
	at org.eclipse.core.internal.events.BuildManager.build(BuildManager.java:381)
	at org.eclipse.core.internal.events.AutoBuildJob.doBuild(AutoBuildJob.java:143)
	at org.eclipse.core.internal.events.AutoBuildJob.run(AutoBuildJob.java:241)
	at org.eclipse.core.internal.jobs.Worker.run(Worker.java:53)
java.lang.NullPointerException
	at com.ge.research.sadl.model.gp.Equation.toFullyQualifiedString(Equation.java:256)
	at com.ge.research.sadl.jena.JenaBasedSadlModelProcessor.createExternalEquation(JenaBasedSadlModelProcessor.java:2968)
	at com.ge.research.sadl.jena.JenaBasedSadlModelProcessor.processStatement(JenaBasedSadlModelProcessor.java:2895)
	at com.ge.research.sadl.jena.JenaBasedSadlModelProcessor.processModelElement(JenaBasedSadlModelProcessor.java:1194)
	at com.ge.research.sadl.jena.JenaBasedSadlModelProcessor.onValidate(JenaBasedSadlModelProcessor.java:1078)
	at com.ge.research.sadl.validation.ResourceValidator.doValidate(ResourceValidator.java:76)
	at com.ge.research.sadl.validation.ResourceValidator.validate(ResourceValidator.java:72)
	at org.eclipse.xtext.validation.ResourceValidatorImpl.validate(ResourceValidatorImpl.java:90)
	at com.ge.research.sadl.validation.ResourceValidator.lambda$0(ResourceValidator.java:56)
	at org.eclipse.xtext.util.OnChangeEvictingCache.get(OnChangeEvictingCache.java:76)
	at com.ge.research.sadl.validation.ResourceValidator.validate(ResourceValidator.java:58)
	at org.eclipse.xtext.ui.editor.validation.ValidationJob$1.exec(ValidationJob.java:91)
	at org.eclipse.xtext.ui.editor.validation.ValidationJob$1.exec(ValidationJob.java:1)
	at org.eclipse.xtext.util.concurrent.CancelableUnitOfWork.exec(CancelableUnitOfWork.java:26)
	at org.eclipse.xtext.resource.OutdatedStateManager.exec(OutdatedStateManager.java:91)
	at org.eclipse.xtext.ui.editor.model.XtextDocument$XtextDocumentLocker.internalReadOnly(XtextDocument.java:527)
	at org.eclipse.xtext.ui.editor.model.XtextDocument$XtextDocumentLocker.readOnly(XtextDocument.java:499)
	at org.eclipse.xtext.ui.editor.model.XtextDocument.readOnly(XtextDocument.java:137)
	at org.eclipse.xtext.ui.editor.validation.ValidationJob.createIssues(ValidationJob.java:86)
	at org.eclipse.xtext.ui.editor.validation.ValidationJob.run(ValidationJob.java:67)
	at org.eclipse.core.internal.jobs.Worker.run(Worker.java:53)
java.lang.NullPointerException
	at com.ge.research.sadl.model.gp.Equation.toFullyQualifiedString(Equation.java:256)
	at com.ge.research.sadl.jena.JenaBasedSadlModelProcessor.createExternalEquation(JenaBasedSadlModelProcessor.java:2968)
	at com.ge.research.sadl.jena.JenaBasedSadlModelProcessor.processStatement(JenaBasedSadlModelProcessor.java:2895)
	at com.ge.research.sadl.jena.JenaBasedSadlModelProcessor.processModelElement(JenaBasedSadlModelProcessor.java:1194)
	at com.ge.research.sadl.jena.JenaBasedSadlModelProcessor.onValidate(JenaBasedSadlModelProcessor.java:1078)
	at com.ge.research.sadl.validation.ResourceValidator.doValidate(ResourceValidator.java:76)
	at com.ge.research.sadl.validation.ResourceValidator.validate(ResourceValidator.java:72)
	at org.eclipse.xtext.validation.ResourceValidatorImpl.validate(ResourceValidatorImpl.java:90)
	at com.ge.research.sadl.validation.ResourceValidator.lambda$0(ResourceValidator.java:56)
	at org.eclipse.xtext.util.OnChangeEvictingCache.get(OnChangeEvictingCache.java:76)
	at com.ge.research.sadl.validation.ResourceValidator.validate(ResourceValidator.java:58)
	at org.eclipse.xtext.ui.validation.DefaultResourceUIValidatorExtension.addMarkers(DefaultResourceUIValidatorExtension.java:60)
	at org.eclipse.xtext.ui.validation.DefaultResourceUIValidatorExtension.updateValidationMarkers(DefaultResourceUIValidatorExtension.java:46)
	at org.eclipse.xtext.builder.builderState.MarkerUpdaterImpl.processDelta(MarkerUpdaterImpl.java:93)
	at org.eclipse.xtext.builder.builderState.MarkerUpdaterImpl.updateMarkers(MarkerUpdaterImpl.java:63)
	at org.eclipse.xtext.builder.builderState.AbstractBuilderState.updateMarkers(AbstractBuilderState.java:82)
	at org.eclipse.xtext.builder.clustering.ClusteringBuilderState.doUpdate(ClusteringBuilderState.java:283)
	at org.eclipse.xtext.builder.builderState.AbstractBuilderState.update(AbstractBuilderState.java:116)
	at org.eclipse.xtext.builder.impl.XtextBuilder.doBuild(XtextBuilder.java:288)
	at org.eclipse.xtext.builder.impl.XtextBuilder.incrementalBuild(XtextBuilder.java:268)
	at org.eclipse.xtext.builder.impl.XtextBuilder.build(XtextBuilder.java:162)
	at org.eclipse.core.internal.events.BuildManager$2.run(BuildManager.java:733)
	at org.eclipse.core.runtime.SafeRunner.run(SafeRunner.java:42)
	at org.eclipse.core.internal.events.BuildManager.basicBuild(BuildManager.java:206)
	at org.eclipse.core.internal.events.BuildManager.basicBuild(BuildManager.java:246)
	at org.eclipse.core.internal.events.BuildManager$1.run(BuildManager.java:299)
	at org.eclipse.core.runtime.SafeRunner.run(SafeRunner.java:42)
	at org.eclipse.core.internal.events.BuildManager.basicBuild(BuildManager.java:302)
	at org.eclipse.core.internal.events.BuildManager.basicBuildLoop(BuildManager.java:358)
	at org.eclipse.core.internal.events.BuildManager.build(BuildManager.java:381)
	at org.eclipse.core.internal.events.AutoBuildJob.doBuild(AutoBuildJob.java:143)
	at org.eclipse.core.internal.events.AutoBuildJob.run(AutoBuildJob.java:241)
	at org.eclipse.core.internal.jobs.Worker.run(Worker.java:53)
java.lang.NullPointerException
	at com.ge.research.sadl.model.gp.Equation.toFullyQualifiedString(Equation.java:256)
	at com.ge.research.sadl.jena.JenaBasedSadlModelProcessor.createExternalEquation(JenaBasedSadlModelProcessor.java:2968)
	at com.ge.research.sadl.jena.JenaBasedSadlModelProcessor.processStatement(JenaBasedSadlModelProcessor.java:2895)
	at com.ge.research.sadl.jena.JenaBasedSadlModelProcessor.processModelElement(JenaBasedSadlModelProcessor.java:1194)
	at com.ge.research.sadl.jena.JenaBasedSadlModelProcessor.onValidate(JenaBasedSadlModelProcessor.java:1078)
	at com.ge.research.sadl.validation.ResourceValidator.doValidate(ResourceValidator.java:76)
	at com.ge.research.sadl.validation.ResourceValidator.validate(ResourceValidator.java:72)
	at org.eclipse.xtext.validation.ResourceValidatorImpl.validate(ResourceValidatorImpl.java:90)
	at com.ge.research.sadl.validation.ResourceValidator.lambda$0(ResourceValidator.java:56)
	at org.eclipse.xtext.util.OnChangeEvictingCache.get(OnChangeEvictingCache.java:76)
	at com.ge.research.sadl.validation.ResourceValidator.validate(ResourceValidator.java:58)
	at org.eclipse.xtext.ui.editor.validation.ValidationJob$1.exec(ValidationJob.java:91)
	at org.eclipse.xtext.ui.editor.validation.ValidationJob$1.exec(ValidationJob.java:1)
	at org.eclipse.xtext.util.concurrent.CancelableUnitOfWork.exec(CancelableUnitOfWork.java:26)
	at org.eclipse.xtext.resource.OutdatedStateManager.exec(OutdatedStateManager.java:91)
	at org.eclipse.xtext.ui.editor.model.XtextDocument$XtextDocumentLocker.internalReadOnly(XtextDocument.java:527)
	at org.eclipse.xtext.ui.editor.model.XtextDocument$XtextDocumentLocker.readOnly(XtextDocument.java:499)
	at org.eclipse.xtext.ui.editor.model.XtextDocument.readOnly(XtextDocument.java:137)
	at org.eclipse.xtext.ui.editor.validation.ValidationJob.createIssues(ValidationJob.java:86)
	at org.eclipse.xtext.ui.editor.validation.ValidationJob.run(ValidationJob.java:67)
	at org.eclipse.core.internal.jobs.Worker.run(Worker.java:53)
java.lang.NullPointerException
	at com.ge.research.sadl.model.gp.Equation.toFullyQualifiedString(Equation.java:256)
	at com.ge.research.sadl.jena.JenaBasedSadlModelProcessor.createExternalEquation(JenaBasedSadlModelProcessor.java:2968)
	at com.ge.research.sadl.jena.JenaBasedSadlModelProcessor.processStatement(JenaBasedSadlModelProcessor.java:2895)
	at com.ge.research.sadl.jena.JenaBasedSadlModelProcessor.processModelElement(JenaBasedSadlModelProcessor.java:1194)
	at com.ge.research.sadl.jena.JenaBasedSadlModelProcessor.onValidate(JenaBasedSadlModelProcessor.java:1078)
	at com.ge.research.sadl.validation.ResourceValidator.doValidate(ResourceValidator.java:76)
	at com.ge.research.sadl.validation.ResourceValidator.validate(ResourceValidator.java:72)
	at org.eclipse.xtext.validation.ResourceValidatorImpl.validate(ResourceValidatorImpl.java:90)
	at com.ge.research.sadl.validation.ResourceValidator.lambda$0(ResourceValidator.java:56)
	at org.eclipse.xtext.util.OnChangeEvictingCache.get(OnChangeEvictingCache.java:76)
	at com.ge.research.sadl.validation.ResourceValidator.validate(ResourceValidator.java:58)
	at org.eclipse.xtext.ui.editor.validation.ValidationJob$1.exec(ValidationJob.java:91)
	at org.eclipse.xtext.ui.editor.validation.ValidationJob$1.exec(ValidationJob.java:1)
	at org.eclipse.xtext.util.concurrent.CancelableUnitOfWork.exec(CancelableUnitOfWork.java:26)
	at org.eclipse.xtext.resource.OutdatedStateManager.exec(OutdatedStateManager.java:91)
	at org.eclipse.xtext.ui.editor.model.XtextDocument$XtextDocumentLocker.internalReadOnly(XtextDocument.java:527)
	at org.eclipse.xtext.ui.editor.model.XtextDocument$XtextDocumentLocker.readOnly(XtextDocument.java:499)
	at org.eclipse.xtext.ui.editor.model.XtextDocument.readOnly(XtextDocument.java:137)
	at org.eclipse.xtext.ui.editor.validation.ValidationJob.createIssues(ValidationJob.java:86)
	at org.eclipse.xtext.ui.editor.validation.ValidationJob.run(ValidationJob.java:67)
	at org.eclipse.core.internal.jobs.Worker.run(Worker.java:53)
java.lang.NullPointerException
	at com.ge.research.sadl.model.gp.Equation.toFullyQualifiedString(Equation.java:256)
	at com.ge.research.sadl.jena.JenaBasedSadlModelProcessor.createExternalEquation(JenaBasedSadlModelProcessor.java:2968)
	at com.ge.research.sadl.jena.JenaBasedSadlModelProcessor.processStatement(JenaBasedSadlModelProcessor.java:2895)
	at com.ge.research.sadl.jena.JenaBasedSadlModelProcessor.processModelElement(JenaBasedSadlModelProcessor.java:1194)
	at com.ge.research.sadl.jena.JenaBasedSadlModelProcessor.onValidate(JenaBasedSadlModelProcessor.java:1078)
	at com.ge.research.sadl.validation.ResourceValidator.doValidate(ResourceValidator.java:76)
	at com.ge.research.sadl.validation.ResourceValidator.validate(ResourceValidator.java:72)
	at org.eclipse.xtext.validation.ResourceValidatorImpl.validate(ResourceValidatorImpl.java:90)
	at com.ge.research.sadl.validation.ResourceValidator.lambda$0(ResourceValidator.java:56)
	at org.eclipse.xtext.util.OnChangeEvictingCache.get(OnChangeEvictingCache.java:76)
	at com.ge.research.sadl.validation.ResourceValidator.validate(ResourceValidator.java:58)
	at org.eclipse.xtext.ui.validation.DefaultResourceUIValidatorExtension.addMarkers(DefaultResourceUIValidatorExtension.java:60)
	at org.eclipse.xtext.ui.validation.DefaultResourceUIValidatorExtension.updateValidationMarkers(DefaultResourceUIValidatorExtension.java:46)
	at org.eclipse.xtext.builder.builderState.MarkerUpdaterImpl.processDelta(MarkerUpdaterImpl.java:93)
	at org.eclipse.xtext.builder.builderState.MarkerUpdaterImpl.updateMarkers(MarkerUpdaterImpl.java:63)
	at org.eclipse.xtext.builder.builderState.AbstractBuilderState.updateMarkers(AbstractBuilderState.java:82)
	at org.eclipse.xtext.builder.clustering.ClusteringBuilderState.doUpdate(ClusteringBuilderState.java:283)
	at org.eclipse.xtext.builder.builderState.AbstractBuilderState.update(AbstractBuilderState.java:116)
	at org.eclipse.xtext.builder.impl.XtextBuilder.doBuild(XtextBuilder.java:288)
	at org.eclipse.xtext.builder.impl.XtextBuilder.incrementalBuild(XtextBuilder.java:268)
	at org.eclipse.xtext.builder.impl.XtextBuilder.build(XtextBuilder.java:162)
	at org.eclipse.core.internal.events.BuildManager$2.run(BuildManager.java:733)
	at org.eclipse.core.runtime.SafeRunner.run(SafeRunner.java:42)
	at org.eclipse.core.internal.events.BuildManager.basicBuild(BuildManager.java:206)
	at org.eclipse.core.internal.events.BuildManager.basicBuild(BuildManager.java:246)
	at org.eclipse.core.internal.events.BuildManager$1.run(BuildManager.java:299)
	at org.eclipse.core.runtime.SafeRunner.run(SafeRunner.java:42)
	at org.eclipse.core.internal.events.BuildManager.basicBuild(BuildManager.java:302)
	at org.eclipse.core.internal.events.BuildManager.basicBuildLoop(BuildManager.java:358)
	at org.eclipse.core.internal.events.BuildManager.build(BuildManager.java:381)
	at org.eclipse.core.internal.events.AutoBuildJob.doBuild(AutoBuildJob.java:143)
	at org.eclipse.core.internal.events.AutoBuildJob.run(AutoBuildJob.java:241)
	at org.eclipse.core.internal.jobs.Worker.run(Worker.java:53)
java.lang.NullPointerException
	at com.ge.research.sadl.model.gp.Equation.toFullyQualifiedString(Equation.java:256)
	at com.ge.research.sadl.jena.JenaBasedSadlModelProcessor.createExternalEquation(JenaBasedSadlModelProcessor.java:2968)
	at com.ge.research.sadl.jena.JenaBasedSadlModelProcessor.processStatement(JenaBasedSadlModelProcessor.java:2895)
	at com.ge.research.sadl.jena.JenaBasedSadlModelProcessor.processModelElement(JenaBasedSadlModelProcessor.java:1194)
	at com.ge.research.sadl.jena.JenaBasedSadlModelProcessor.onValidate(JenaBasedSadlModelProcessor.java:1078)
	at com.ge.research.sadl.validation.ResourceValidator.doValidate(ResourceValidator.java:76)
	at com.ge.research.sadl.validation.ResourceValidator.validate(ResourceValidator.java:72)
	at org.eclipse.xtext.validation.ResourceValidatorImpl.validate(ResourceValidatorImpl.java:90)
	at com.ge.research.sadl.validation.ResourceValidator.lambda$0(ResourceValidator.java:56)
	at org.eclipse.xtext.util.OnChangeEvictingCache.get(OnChangeEvictingCache.java:76)
	at com.ge.research.sadl.validation.ResourceValidator.validate(ResourceValidator.java:58)
	at org.eclipse.xtext.ui.editor.validation.ValidationJob$1.exec(ValidationJob.java:91)
	at org.eclipse.xtext.ui.editor.validation.ValidationJob$1.exec(ValidationJob.java:1)
	at org.eclipse.xtext.util.concurrent.CancelableUnitOfWork.exec(CancelableUnitOfWork.java:26)
	at org.eclipse.xtext.resource.OutdatedStateManager.exec(OutdatedStateManager.java:91)
	at org.eclipse.xtext.ui.editor.model.XtextDocument$XtextDocumentLocker.internalReadOnly(XtextDocument.java:527)
	at org.eclipse.xtext.ui.editor.model.XtextDocument$XtextDocumentLocker.readOnly(XtextDocument.java:499)
	at org.eclipse.xtext.ui.editor.model.XtextDocument.readOnly(XtextDocument.java:137)
	at org.eclipse.xtext.ui.editor.validation.ValidationJob.createIssues(ValidationJob.java:86)
	at org.eclipse.xtext.ui.editor.validation.ValidationJob.run(ValidationJob.java:67)
	at org.eclipse.core.internal.jobs.Worker.run(Worker.java:53)
java.lang.NullPointerException
	at com.ge.research.sadl.model.gp.Equation.toFullyQualifiedString(Equation.java:256)
	at com.ge.research.sadl.jena.JenaBasedSadlModelProcessor.createExternalEquation(JenaBasedSadlModelProcessor.java:2968)
	at com.ge.research.sadl.jena.JenaBasedSadlModelProcessor.processStatement(JenaBasedSadlModelProcessor.java:2895)
	at com.ge.research.sadl.jena.JenaBasedSadlModelProcessor.processModelElement(JenaBasedSadlModelProcessor.java:1194)
	at com.ge.research.sadl.jena.JenaBasedSadlModelProcessor.onValidate(JenaBasedSadlModelProcessor.java:1078)
	at com.ge.research.sadl.validation.ResourceValidator.doValidate(ResourceValidator.java:76)
	at com.ge.research.sadl.validation.ResourceValidator.validate(ResourceValidator.java:72)
	at org.eclipse.xtext.validation.ResourceValidatorImpl.validate(ResourceValidatorImpl.java:90)
	at com.ge.research.sadl.validation.ResourceValidator.lambda$0(ResourceValidator.java:56)
	at org.eclipse.xtext.util.OnChangeEvictingCache.get(OnChangeEvictingCache.java:76)
	at com.ge.research.sadl.validation.ResourceValidator.validate(ResourceValidator.java:58)
	at org.eclipse.xtext.ui.validation.DefaultResourceUIValidatorExtension.addMarkers(DefaultResourceUIValidatorExtension.java:60)
	at org.eclipse.xtext.ui.validation.DefaultResourceUIValidatorExtension.updateValidationMarkers(DefaultResourceUIValidatorExtension.java:46)
	at org.eclipse.xtext.builder.builderState.MarkerUpdaterImpl.processDelta(MarkerUpdaterImpl.java:93)
	at org.eclipse.xtext.builder.builderState.MarkerUpdaterImpl.updateMarkers(MarkerUpdaterImpl.java:63)
	at org.eclipse.xtext.builder.builderState.AbstractBuilderState.updateMarkers(AbstractBuilderState.java:82)
	at org.eclipse.xtext.builder.clustering.ClusteringBuilderState.doUpdate(ClusteringBuilderState.java:283)
	at org.eclipse.xtext.builder.builderState.AbstractBuilderState.update(AbstractBuilderState.java:116)
	at org.eclipse.xtext.builder.impl.XtextBuilder.doBuild(XtextBuilder.java:288)
	at org.eclipse.xtext.builder.impl.XtextBuilder.incrementalBuild(XtextBuilder.java:268)
	at org.eclipse.xtext.builder.impl.XtextBuilder.build(XtextBuilder.java:162)
	at org.eclipse.core.internal.events.BuildManager$2.run(BuildManager.java:733)
	at org.eclipse.core.runtime.SafeRunner.run(SafeRunner.java:42)
	at org.eclipse.core.internal.events.BuildManager.basicBuild(BuildManager.java:206)
	at org.eclipse.core.internal.events.BuildManager.basicBuild(BuildManager.java:246)
	at org.eclipse.core.internal.events.BuildManager$1.run(BuildManager.java:299)
	at org.eclipse.core.runtime.SafeRunner.run(SafeRunner.java:42)
	at org.eclipse.core.internal.events.BuildManager.basicBuild(BuildManager.java:302)
	at org.eclipse.core.internal.events.BuildManager.basicBuildLoop(BuildManager.java:358)
	at org.eclipse.core.internal.events.BuildManager.build(BuildManager.java:381)
	at org.eclipse.core.internal.events.AutoBuildJob.doBuild(AutoBuildJob.java:143)
	at org.eclipse.core.internal.events.AutoBuildJob.run(AutoBuildJob.java:241)
	at org.eclipse.core.internal.jobs.Worker.run(Worker.java:53)
java.lang.NullPointerException
	at com.ge.research.sadl.model.gp.Equation.toFullyQualifiedString(Equation.java:256)
	at com.ge.research.sadl.jena.JenaBasedSadlModelProcessor.createExternalEquation(JenaBasedSadlModelProcessor.java:2968)
	at com.ge.research.sadl.jena.JenaBasedSadlModelProcessor.processStatement(JenaBasedSadlModelProcessor.java:2895)
	at com.ge.research.sadl.jena.JenaBasedSadlModelProcessor.processModelElement(JenaBasedSadlModelProcessor.java:1194)
	at com.ge.research.sadl.jena.JenaBasedSadlModelProcessor.onValidate(JenaBasedSadlModelProcessor.java:1078)
	at com.ge.research.sadl.validation.ResourceValidator.doValidate(ResourceValidator.java:76)
	at com.ge.research.sadl.validation.ResourceValidator.validate(ResourceValidator.java:72)
	at org.eclipse.xtext.validation.ResourceValidatorImpl.validate(ResourceValidatorImpl.java:90)
	at com.ge.research.sadl.validation.ResourceValidator.lambda$0(ResourceValidator.java:56)
	at org.eclipse.xtext.util.OnChangeEvictingCache.get(OnChangeEvictingCache.java:76)
	at com.ge.research.sadl.validation.ResourceValidator.validate(ResourceValidator.java:58)
	at org.eclipse.xtext.ui.editor.validation.ValidationJob$1.exec(ValidationJob.java:91)
	at org.eclipse.xtext.ui.editor.validation.ValidationJob$1.exec(ValidationJob.java:1)
	at org.eclipse.xtext.util.concurrent.CancelableUnitOfWork.exec(CancelableUnitOfWork.java:26)
	at org.eclipse.xtext.resource.OutdatedStateManager.exec(OutdatedStateManager.java:91)
	at org.eclipse.xtext.ui.editor.model.XtextDocument$XtextDocumentLocker.internalReadOnly(XtextDocument.java:527)
	at org.eclipse.xtext.ui.editor.model.XtextDocument$XtextDocumentLocker.readOnly(XtextDocument.java:499)
	at org.eclipse.xtext.ui.editor.model.XtextDocument.readOnly(XtextDocument.java:137)
	at org.eclipse.xtext.ui.editor.validation.ValidationJob.createIssues(ValidationJob.java:86)
	at org.eclipse.xtext.ui.editor.validation.ValidationJob.run(ValidationJob.java:67)
	at org.eclipse.core.internal.jobs.Worker.run(Worker.java:53)
```

--- a/sadl3/com.ge.research.sadl.parent/com.ge.research.sadl.tests/src/com/ge/research/sadl/tests/AbstractLinkingTest.xtend
+++ b/sadl3/com.ge.research.sadl.parent/com.ge.research.sadl.tests/src/com/ge/research/sadl/tests/AbstractLinkingTest.xtend
@@ -38,6 +38,15 @@ import com.ge.research.sadl.sADL.QueryStatement
 import com.ge.research.sadl.tests.helpers.XtendTemplateHelper
 import com.ge.research.sadl.sADL.SadlRangeRestriction
 
+/**
+ * Notes on writing linking tests:
+ * <p>
+ * <ul>
+ * <li>When defining the SADL model, <b>[RESOURCE]</b> is the <b>definition</b> of a SADL resource and <i>&lt;RESOURCE&gt;</i> is the <i>reference</i>.</li>
+ * <li>When CTRL/CMD + Clicking on an item, <b>[RESORUCE]</b> stays, as it is the <b>definition</b> and <i>&lt;RESOURCE&gt;</i> jumps to <b>[RESORUCE]</b>.</li>
+ * </ul>
+ * </p>
+ */
 @RunWith(XtextRunner)
 @InjectWith(SADLInjectorProvider)
 abstract class AbstractLinkingTest extends AbstractSadlTest {
@@ -89,8 +98,14 @@ abstract class AbstractLinkingTest extends AbstractSadlTest {
 			var result = parseableContents
 			var addedChars = 0
 			for (decl : actualNames.sortBy[region.offset]) {
-				val brackets = if (decl.isDeclaration === null) "?"->"?" else if (decl.isDeclaration) "["->"]" else "<"->">"
-				result = result.substring(0,decl.region.offset + addedChars) + brackets.key+ decl.value + brackets.value + result.substring(decl.region.offset + decl.region.length + addedChars)
+				val brackets = if (decl.isDeclaration === null) {
+					"?"->"?" 
+				} else if (decl.isDeclaration) {
+					"["->"]"
+				} else {
+					"<"->">"
+				}
+				result = result.substring(0, decl.region.offset + addedChars) + brackets.key + decl.value + brackets.value + result.substring(decl.region.offset + decl.region.length + addedChars)
 				addedChars += 2
 			}
 			return result
@@ -103,8 +118,8 @@ abstract class AbstractLinkingTest extends AbstractSadlTest {
 		TextRegion region
 	}
 	
-	private static val NAME = "[a-zA-Z0-9_\\.:]+"
-	private static val markerPattern = Pattern.compile("[\\[\\<]("+NAME+")[\\]|\\>]")
+	static val NAME = "[a-zA-Z0-9_\\.:]+"
+	static val markerPattern = Pattern.compile("[\\[\\<]("+NAME+")[\\]|\\>]")
 	
 	private def TestFile parseReferenceMarker(CharSequence txt) {
 		val result = new TestFile() => [

--- a/sadl3/com.ge.research.sadl.parent/com.ge.research.sadl.tests/src/com/ge/research/sadl/tests/scoping/SadlLinkingTest.xtend
+++ b/sadl3/com.ge.research.sadl.parent/com.ge.research.sadl.tests/src/com/ge/research/sadl/tests/scoping/SadlLinkingTest.xtend
@@ -1340,35 +1340,44 @@ class SadlLinkingTest extends AbstractLinkingTest {
 		B is a type of <seconds>.
 		'''.assertLinking[sadl]
 	}
-	
+
+	/**
+	 * From: https://github.com/crapo/sadlos2/issues/344#issue-382432616
+	 * <ul>
+	 * <li><b>a)</b> If a variable is defined in the "where" clause and used in the body, the definition is the reference in the "where" clause.</li>
+	 * <li><b>b)</b> If a variable is referenced in the body and referenced in the "return" statement, the definition is the reference in the body.</li>
+	 * </ul>
+	 */
 	@Test
-	def void testEquationStatementScoping_GH344() {
+	def void testEquationStatementScoping_GH_344_A() {
 		'''
+			// ------ SETUP ------
 			uri "http://sadl.org/MinimalExample.sadl" alias mex.
-			
 			ScientificConcept is a class.
 			sadlimplicitmodel:UnittedQuantity is a type of ScientificConcept.
-			
 			External unitResolver(string u, ...) returns string: "http://sadl.org/unitSelector".
 			External derivative(ScientificConcept numerator, ScientificConcept denominator, int n) returns decimal, string: "http://sadl.org/derivative".
-			
 			Mass is a type of UnittedQuantity.
 			Velocity is a type of UnittedQuantity.
 			Momentum is a type of UnittedQuantity.
 			Force is a type of UnittedQuantity.
 			Time is a type of UnittedQuantity.
+			// ------ END OF SETUP ------
 			
 			Equation newtons2ndLaw
 			  (note "Force is equal to the derivative of momentum with respect to time.")
-			  (Mass m, Velocity v) returns Force:
-			  a Force f with ^value fv, with unit fu
+			  (Mass [m], Velocity [v]) returns Force:
+			// BODY
+			  a Force f with ^value <fv>, with unit <fu>
 			
-			  return f
+			// RETURN
+			  return <f>
 			
-			  where mv is a Momentum
-			    with ^value (^value of m * ^value of v),
-			    with unit unitResolver("*", unit of m, unit of v)
-			    and [fv,fu] = derivative(mv, Time, 1).
+			// WHERE
+			  where [mv] is a Momentum
+			    with ^value (^value of <m> * ^value of <v>),
+			    with unit unitResolver("*", unit of <m>, unit of <v>)
+			    and [[fv],[fu]] = derivative(<mv>, Time, 1).
 		'''.assertLinking[sadl];
 	}
 }

--- a/sadl3/com.ge.research.sadl.parent/com.ge.research.sadl.tests/src/com/ge/research/sadl/tests/scoping/SadlLinkingTest.xtend
+++ b/sadl3/com.ge.research.sadl.parent/com.ge.research.sadl.tests/src/com/ge/research/sadl/tests/scoping/SadlLinkingTest.xtend
@@ -1340,4 +1340,35 @@ class SadlLinkingTest extends AbstractLinkingTest {
 		B is a type of <seconds>.
 		'''.assertLinking[sadl]
 	}
+	
+	@Test
+	def void testEquationStatementScoping_GH344() {
+		'''
+			uri "http://sadl.org/MinimalExample.sadl" alias mex.
+			
+			ScientificConcept is a class.
+			sadlimplicitmodel:UnittedQuantity is a type of ScientificConcept.
+			
+			External unitResolver(string u, ...) returns string: "http://sadl.org/unitSelector".
+			External derivative(ScientificConcept numerator, ScientificConcept denominator, int n) returns decimal, string: "http://sadl.org/derivative".
+			
+			Mass is a type of UnittedQuantity.
+			Velocity is a type of UnittedQuantity.
+			Momentum is a type of UnittedQuantity.
+			Force is a type of UnittedQuantity.
+			Time is a type of UnittedQuantity.
+			
+			Equation newtons2ndLaw
+			  (note "Force is equal to the derivative of momentum with respect to time.")
+			  (Mass m, Velocity v) returns Force:
+			  a Force f with ^value fv, with unit fu
+			
+			  return f
+			
+			  where mv is a Momentum
+			    with ^value (^value of m * ^value of v),
+			    with unit unitResolver("*", unit of m, unit of v)
+			    and [fv,fu] = derivative(mv, Time, 1).
+		'''.assertLinking[sadl];
+	}
 }

--- a/sadl3/com.ge.research.sadl.parent/com.ge.research.sadl/src/com/ge/research/sadl/scoping/SADLScopeProvider.xtend
+++ b/sadl3/com.ge.research.sadl.parent/com.ge.research.sadl/src/com/ge/research/sadl/scoping/SADLScopeProvider.xtend
@@ -43,10 +43,13 @@ import com.ge.research.sadl.sADL.SadlResource
 import com.ge.research.sadl.sADL.SelectExpression
 import com.ge.research.sadl.sADL.SubjHasProp
 import com.ge.research.sadl.sADL.TestStatement
+import com.ge.research.sadl.sADL.UpdateExpression
+import com.ge.research.sadl.sADL.UpdateStatement
 import com.google.common.base.Predicate
 import com.google.common.base.Predicates
 import com.google.common.collect.ImmutableMap
 import com.google.inject.Inject
+import com.google.inject.Provider
 import java.util.List
 import java.util.Map
 import java.util.Set
@@ -71,9 +74,6 @@ import static com.ge.research.sadl.processing.SadlConstants.*
 import static com.ge.research.sadl.sADL.SADLPackage.Literals.*
 
 import static extension com.ge.research.sadl.utils.SadlASTUtils.*
-import com.ge.research.sadl.sADL.UpdateStatement
-import com.ge.research.sadl.sADL.UpdateExpression
-import com.google.inject.Provider
 
 /**
  * This class contains custom scoping description.
@@ -93,14 +93,14 @@ class SADLScopeProvider extends AbstractGlobalScopeDelegatingScopeProvider {
 	 */	
 	static val IMPLICIT_MODELS = ImmutableMap.of(SADL_IMPLICIT_MODEL_FILENAME, SADL_IMPLICIT_MODEL_URI,
 		SADL_BUILTIN_FUNCTIONS_FILENAME, IReasoner.SADL_BUILTIN_FUNCTIONS_URI);
-		
+
 	@Inject protected extension DeclarationExtensions
 	@Inject protected IQualifiedNameConverter converter
 	@Inject protected AmbiguousNameHelper ambiguousNameHelper;
 	@Inject protected GlobalScopeProviderFilterProvider globalScopeProviderFilterProvider;
-	
+
 	@Inject OnChangeEvictingCache cache
-	
+
 	val LocalScopeProvider localScope_01 = namedScopeProvider([resource, namespace, parentScope, importScope |
 		return internalGetLocalResourceScope(resource, namespace, parentScope, importScope, true) [
 			if (it instanceof SadlResource) {
@@ -111,7 +111,7 @@ class SADLScopeProvider extends AbstractGlobalScopeDelegatingScopeProvider {
 			return false;
 		];
 	], '01');
-	
+
 	val LocalScopeProvider localScope_02 = namedScopeProvider([resource, namespace, parentScope, importScope |
 		return internalGetLocalResourceScope(resource, namespace, parentScope, importScope) [
 			if (it instanceof SadlResource) {
@@ -131,7 +131,7 @@ class SADLScopeProvider extends AbstractGlobalScopeDelegatingScopeProvider {
 			return false;
 		];
 	], '03');
-	
+
 	// SADL nested instance support: https://github.com/crapo/sadlos2/issues/250
 	val LocalScopeProvider localScope_04 = namedScopeProvider([resource, namespace, parentScope, importScope |
 		return internalGetLocalResourceScope(resource, namespace, parentScope, importScope) [
@@ -141,7 +141,7 @@ class SADLScopeProvider extends AbstractGlobalScopeDelegatingScopeProvider {
 			return false;
 		];
 	], '04');
-	
+
 	val LocalScopeProvider localScope_05 = namedScopeProvider([resource, namespace, parentScope, importScope |
 		return internalGetLocalResourceScope(resource, namespace, parentScope, importScope) [
 			if (it instanceof SadlResource) {
@@ -160,8 +160,20 @@ class SADLScopeProvider extends AbstractGlobalScopeDelegatingScopeProvider {
 			return false;
 		];
 	], '06');
-	
-	
+
+	// https://github.com/crapo/sadlos2/issues/344
+	val LocalScopeProvider localScope_07 = namedScopeProvider([resource, namespace, parentScope, importScope |
+		return internalGetLocalResourceScope(resource, namespace, parentScope, importScope) [
+			if (it instanceof SadlResource) {
+				val equationStatement = EcoreUtil2.getContainerOfType(it, EquationStatement);
+				if (equationStatement !== null) {
+					return equationStatement.where !== null && EcoreUtil2.isAncestor(equationStatement.where, it);
+				}
+			} 
+			return false;
+		];
+	], '07');
+
 	override getScope(EObject context, EReference reference) {
 		// resolving imports against external models goes directly to the global scope
 		if (reference.EReferenceType === SADL_MODEL) {
@@ -178,7 +190,7 @@ class SADLScopeProvider extends AbstractGlobalScopeDelegatingScopeProvider {
 		throw new UnsupportedOperationException(
 			"Couldn't build scope for elements of type " + reference.EReferenceType.name)
 	}
-	
+
 	protected def IScope getSadlResourceScope(EObject context, EReference reference) {
 		val parent = createResourceScope(context.eResource, null, newHashSet)
 		
@@ -188,8 +200,19 @@ class SADLScopeProvider extends AbstractGlobalScopeDelegatingScopeProvider {
 		}
 		val equation = EcoreUtil2.getContainerOfType(context, EquationStatement)
 		if (equation !== null) {
-			return MapBasedScope.createScope(parent, 
-				equation.parameter.map[EObjectDescription.create(name.concreteName, it.name)])
+			// Filter items with `null` `concreteName`.
+			// E.g.: `Equation add(decimal x, /* no "type" here */ y) returns decimal: x + y.`
+			val parentScopeForParameters = MapBasedScope.createScope(parent,
+				equation.parameter
+					.map[name.concreteName -> it]
+					.filter[!key.nullOrEmpty]
+					.map[EObjectDescription.create(key, value.name)])
+
+			// Equation can have `body`, `where`, and `retval`. (https://github.com/crapo/sadlos2/issues/344#issue-382432616)
+			// So we cannot just stop after adding the SADL resources from the parameter list to the scope,
+			// but we have to process them further for the local scope.
+			val childExpressions = #[equation.body, equation.where, equation.retval].filterNull
+			return getLocalVariableScope(childExpressions, parentScopeForParameters)
 		}
 		val ask = EcoreUtil2.getContainerOfType(context, QueryStatement)
 		if (ask?.expr !== null) {
@@ -206,7 +229,7 @@ class SADLScopeProvider extends AbstractGlobalScopeDelegatingScopeProvider {
 		}
 		return parent
 	}
-	
+
 	protected def IScope getLocalVariableScope(Iterable<Expression> expressions, IScope parent) {
 		if (expressions.nullOrEmpty) {
 			return parent;
@@ -225,7 +248,7 @@ class SADLScopeProvider extends AbstractGlobalScopeDelegatingScopeProvider {
 			if (container instanceof BinaryOperation) {
 				if (container.op == 'is' || container.op == '==' || container.op == '=' ||
 					(container.op == 'and' && EcoreUtil2.getContainerOfType(it, SelectExpression) !== null) || // we are in the middle of a select expression.
-					(container.op == 'and' && EcoreUtil2.getContainerOfType(it, RuleStatement) !== null)    // we are in the middle of a rule statement (GH-245).
+					(container.op == 'and' && EcoreUtil2.getContainerOfType(it, RuleStatement) !== null) // we are in the middle of a rule statement (GH-245).
 				) {
 					return true;
 				}
@@ -244,7 +267,7 @@ class SADLScopeProvider extends AbstractGlobalScopeDelegatingScopeProvider {
 		]
 		return doGetLocalVariableScope(expressions, newParent, [true]);
 	}
-	
+
 	protected def IScope doGetLocalVariableScope(Iterable<Expression> expressions, IScope parent, Predicate<SadlResource> predicate) {
 		if (expressions.nullOrEmpty) {
 			return parent;
@@ -265,11 +288,11 @@ class SADLScopeProvider extends AbstractGlobalScopeDelegatingScopeProvider {
 		}
 		return MapBasedScope.createScope(parent, map.values)
 	}
-	
+
 	protected def IScope createResourceScope(Resource resource, String alias, Set<Resource> importedResources) {
 		return createResourceScope(resource, alias, importedResources, Predicates.alwaysTrue); 
 	}
-	
+
 	protected def IScope createResourceScope(Resource resource, String alias, Set<Resource> importedResources, Predicate<EObject> isIncluded) {
 		val provider = [
 			val shouldWrap = importedResources.empty
@@ -279,7 +302,7 @@ class SADLScopeProvider extends AbstractGlobalScopeDelegatingScopeProvider {
 			
 			var newParent = createImportScope(resource, importedResources)
 			if (shouldWrap) {
-				newParent = wrap(newParent)				
+				newParent = wrap(newParent)
 			}
 			val importScope = newParent;
 			val aliasToUse = alias ?: resource.getAlias
@@ -302,7 +325,7 @@ class SADLScopeProvider extends AbstractGlobalScopeDelegatingScopeProvider {
 		val key = 'resource_scope' -> alias;
 		return cache.get(key, resource, provider);
 	}
-	
+
 	protected def List<LocalScopeProvider> getLocalScopeProviders() {
 		return newArrayList(
 			localScope_01,
@@ -310,10 +333,11 @@ class SADLScopeProvider extends AbstractGlobalScopeDelegatingScopeProvider {
 			localScope_03,
 			localScope_04,
 			localScope_05,
-			localScope_06
+			localScope_06,
+			localScope_07
 		);
 	}
-	
+
 	protected def IScope internalGetLocalResourceScope(Resource resource, QualifiedName namespace, IScope parentScope,
 		IScope importScope, Predicate<EObject> isIncluded) {
 
@@ -413,7 +437,7 @@ class SADLScopeProvider extends AbstractGlobalScopeDelegatingScopeProvider {
 		}
 		return MapBasedScope.createScope(parentScope, map.values)
 	}
-		
+
 	protected def handleSadlResource(ScopeContext context) {
 		val it = context.obj as SadlResource;
 		val parentScope = context.parentScope;
@@ -458,30 +482,30 @@ class SADLScopeProvider extends AbstractGlobalScopeDelegatingScopeProvider {
 			map.addElement(name2, it)
 		}
 	}
-	
+
 	protected def QualifiedName getName(SadlResource it) {
 		return getName(it, false);
 	}
-	
+
 	protected def QualifiedName getName(SadlResource it, boolean trimPrefix) {
 		return converter.toQualifiedName(getConcreteName(it, trimPrefix));
 	}
-	
+
 	/**
 	 * Clients may override to handle anything before visiting the contents of the resource for the internal resource scope.
 	 */
 	protected def (ScopeContext)=>void doSwitch(ScopeContext context) {
 		return null;
 	}
-	
+
 	protected def pruneScope(EObject object) {
 		return object instanceof RuleStatement || object instanceof EquationStatement || object instanceof QueryStatement || object instanceof TestStatement || object instanceof ExternalEquationStatement
 	}
-	
+
 	protected def String getAlias(Resource resource) {
 		(resource.contents.head as SadlModel).alias
 	}
-	
+
 	protected def IScope createImportScope(Resource resource, Set<Resource> importedResources) {
 		val ambiguousNameDetection = ambiguousNameHelper.isAmbiguousNameCheckEnabled(resource)
 		val imports = resource.contents.head.eContents.filter(SadlImport).toList.reverseView
@@ -519,7 +543,7 @@ class SADLScopeProvider extends AbstractGlobalScopeDelegatingScopeProvider {
 		
 		return new MapScope(IScope.NULLSCOPE, importedSymbols, false)
 	}
-	
+
 	def private IEObjectDescription checkDuplicate(IEObjectDescription existing, IEObjectDescription other, boolean ambiguousNameDetection) {
 		if (!ambiguousNameDetection || existing === null || other === null ||
 			EcoreUtil.getURI(existing.EObjectOrProxy) == EcoreUtil.getURI(other.EObjectOrProxy)) {
@@ -539,25 +563,31 @@ class SADLScopeProvider extends AbstractGlobalScopeDelegatingScopeProvider {
 		}
 		
 	}
-	
+
 	protected def void addElement(Map<QualifiedName, IEObjectDescription> scope, QualifiedName qn, EObject obj) {
 		if (canAddToScope(scope, qn, obj)) {
 			scope.put(qn, new EObjectDescription(qn, obj, emptyMap));
+		} else {
+			if (qn.toString == 'f') {
+				println()
+			}
 		}
 	}
-	
+
 	protected def boolean canAddToScope(Map<QualifiedName, IEObjectDescription> scope, QualifiedName qn, EObject obj) {
 		var canAdd = true;
 		if (obj instanceof SadlResource) {
-			canAdd = !obj.inExternalOrLocalEquationStatement
-				&& !obj.inExpression
-				&& !obj.inQueryStatement
-				&& !obj.inRuleStatement
-				&& !obj.unit;
+			if (!obj.inEquationWhere) {
+				canAdd = !obj.inExternalOrLocalEquationStatement
+					&& !obj.inExpression
+					&& !obj.inQueryStatement
+					&& !obj.inRuleStatement
+					&& !obj.unit;
+			}
 		}
 		return canAdd && !scope.containsKey(qn);
 	}
-	
+
 	// Do not put parameters of external and local equation statements into the scope.
 	protected def boolean isInExternalOrLocalEquationStatement(SadlResource it) {
 		if (eContainer instanceof SadlParameterDeclaration) {
@@ -567,12 +597,12 @@ class SADLScopeProvider extends AbstractGlobalScopeDelegatingScopeProvider {
 		}
 		return false;
 	}
-	
+
 	// Also filter out resources from the expression of any equations.
 	protected def boolean isInExpression(SadlResource it) {
 		return EcoreUtil2.getContainerOfType(it, BinaryOperation) !== null;
 	}
-	
+
 	// The SADL resource from the use-site should not go into the scope.
 	// In such cases the statement does not have an expression.
 	protected def boolean isInQueryStatement(SadlResource it) {
@@ -581,14 +611,26 @@ class SADLScopeProvider extends AbstractGlobalScopeDelegatingScopeProvider {
 		}
 		return false;
 	}
-	
+
 	protected def boolean isInRuleStatement(SadlResource it) {
 		if (eContainer instanceof RuleStatement) {
 			return (eContainer as RuleStatement).thens === null;
 		}
 		return false;
 	}
-	
+
+	/**
+	 * {@code true} if the argument is contained in an {@link EquationStatement}'s {@link EquationStatement#getWhere() where} expression.
+	 * The containment is not necessarily direct. Otherwise, {@code false}.
+	 */
+	protected def boolean isInEquationWhere(SadlResource it) {
+		val equationStatement = EcoreUtil2.getContainerOfType(it, EquationStatement);
+		if (equationStatement !== null && equationStatement.where !== null) {
+			return EcoreUtil2.isAncestor(equationStatement.where, it);
+		}
+		return false;
+	}
+
 	/**
 	 * Sugar for having a name for the local scope provider. This is only for debugging purposes. 
 	 */
@@ -605,7 +647,7 @@ class SADLScopeProvider extends AbstractGlobalScopeDelegatingScopeProvider {
 			
 		}
 	}
-	
+
 	/**
 	 * Local helper that is used to modify the declaration/reference precedences of the SADL resources inside a SADL model.
 	 */
@@ -626,11 +668,11 @@ class SADLScopeProvider extends AbstractGlobalScopeDelegatingScopeProvider {
 		}
 		
 	}
-	
+
 	//----------------------------------------
 	// Helper things.
 	//----------------------------------------
-	
+
 	/**
 	 * Hack to have a static flag bug no warnings due to dead code.
 	 */
@@ -642,5 +684,5 @@ class SADLScopeProvider extends AbstractGlobalScopeDelegatingScopeProvider {
 	static def boolean isDEBUG() {
 		return DEBUG_FLAG === Debug.TRUE;
 	}
-	
+
 }

--- a/sadl3/com.ge.research.sadl.parent/com.ge.research.sadl/src/com/ge/research/sadl/scoping/SADLScopeProvider.xtend
+++ b/sadl3/com.ge.research.sadl.parent/com.ge.research.sadl/src/com/ge/research/sadl/scoping/SADLScopeProvider.xtend
@@ -567,10 +567,6 @@ class SADLScopeProvider extends AbstractGlobalScopeDelegatingScopeProvider {
 	protected def void addElement(Map<QualifiedName, IEObjectDescription> scope, QualifiedName qn, EObject obj) {
 		if (canAddToScope(scope, qn, obj)) {
 			scope.put(qn, new EObjectDescription(qn, obj, emptyMap));
-		} else {
-			if (qn.toString == 'f') {
-				println()
-			}
 		}
 	}
 

